### DISCUSSION TITLE
Placeholder issue seeking guidance for optimizing model run time on WCOSS2 for RRFS application

### DISCUSSION THREAD

**Discussion by MatthewPyle-NOAA** [ufs-community/ufs-weather-model#3023] [General]
We are looking for guidance on optimizing run-time speed on WCOSS2 for a couple of different node count configurations for a large regional domain.  One is for a small node count (~13 nodes) used in making 1 hour forecasts within the EnKF system (and writing a restart file at the 1 h time).  The other is for a large node count (~110 nodes) used in making 60 hour forecasts with hourly history output (and eventually 15 minute history output for the first ~18 hours), and (eventually) 6 h restart writes.  

Input files and a run script has been collected under /lfs/h2/emc/lam/noscrub/Matthew.Pyle/rrfs_optimization_update/ on Cactus and Dogwood

The job_card.sh_6hfore_fullnode script currently is set up to run a 6 h forecast using a 61 node configuration, but can switch to 76, 97, or 110 node configurations easily. If the initialization time could be reduced (it is several minutes for large node counts), and the integration speed could be improved, it would be a great help.   

The job_card.sh_1hfore_fullnode script runs a 1 h forecast, writing out a restart file at the end.  This general configuration is for the EnKF system, which will either be 30 members run concurrently (each member using 12-15 nodes to run) or in two batches of 15 members (each running on 25-30 nodes).  In either scenario, would ideally have all of these forecasts complete within ~10 minutes.

**Comment by GeorgeVandenberghe-NOAA** [comment]
How much I/O does each member do, particularly input.   What are the file names and how large are they and how fast to they need to get read?   There is evidence the WCOSS2 filesystem itself is being crushed by this input when ensembles are started and if so, the I/O needs redesign.  I am late to this investigation and am aware of FMS work to mitigate it

**Comment by GeorgeVandenberghe-NOAA** [comment]
FMS work is described in https://github.com/NOAA-GFDL/FMS/issues/1322

**Comment by gspetro-NOAA** [comment]
@MatthewPyle-NOAA Is this issue ongoing, or has a solution been found?

**Comment by GeorgeVandenberghe-NOAA** [comment]
"How much I/O does each member do, particularly input."

Whether resolved or not for RRFS, this question needs to be answered in reasonable detail.   It was an EE2 specification for all NCO implementations but was dropped after universal developer sentiment that it was too inconvenient to answer and it's dropping has led to runtime variance in ops, fear of running anything alongside ops on production WCOSS and WCOSS2, and inability to determine if an implementation is safe to attempt so I am dead serious this needs to be reexamined and required for all significant implementations and should not be left to questions asked in forensic analysis of a (preventable) mess!

**Comment by MatthewPyle-NOAA** [comment]
@junwang-noaa Here is an older issue that could be repurposed.  

2025 update - looking for optimizations for large (~154 node) RRFS runs going to 84 h.   Part of the e-mail I sent to @DusanJovic-NOAA yesterday:

I think I have everything you should need under /lfs/h2/emc/lam/noscrub/Matthew.Pyle/rrfs_canned on cactus.

The RRFS workflow currently is pulling this model hash:

repo_url = https://github.com/DusanJovic-NOAA/ufs-weather-model.git
hash = 3411cb

I have it set up to clone in /lfs/h2/emc/lam/noscrub/Matthew.Pyle/rrfs_canned/rrfs-workflow/sorc on cactus:

Doing this command: ./app_build.sh --extrn ufs 
will clone and build it as is done for the RRFS, with the executable ending up in rrfs-workflow/exec/ : 

/lfs/h2/emc/lam/noscrub/Matthew.Pyle/rrfs_canned/job_card.sh_*nodes

are simple run scripts for a small node configuration (52 nodes) and for the planned production 84 h forecast configuration (154 nodes).  The rundir has the needed inputs.

**Comment by DusanJovic-NOAA** [comment]
I added tracing calls to the rrfs v1 code (same tracing we used for gfsv17) and ran a full 84hr experiment with both 154 and 172 nodes. These are full run trace plots:

154 nodes:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/55ce2b2f-15ef-4c83-939d-fe4e1a26250d" />

172 nodes:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/c8e159f0-3a86-4117-a7b3-3534d85642e6" />

and here are zoomed in first ~4-5 minutes:

154:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/702557f0-f127-4c8f-8357-c3b435d799c0" />

172:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/41ef74d5-e7e8-44f2-8517-bc4720cb5124" />



Initialization of 172-nodes run is longer than 154-nodes run. The longest running part of fcst_initailize is the section of the code where restart files are read in, in 154 nodes run it takes about 1m 15 sec, while in 172 nodes run it took ~2m 23sec, noticeably longer.

But regardless of that difference overall initialization does not take that long, the first time step (first ModelAdvance call) runs after about 2.5 minutes. The total run takes ~1hr 53min. So in order to have the entire forecast finish in 90min (1h 30min) we need to save ~20min. Even if we reduce initialization to 0 sec (which is of course impossible) we'll still need at least ~15min. Increasing the number of nodes, will slow down the reading of restarts even further. So we need to find out how to make actual integration faster, maybe some physics or dynamic configuration options.

**Comment by junwang-noaa** [comment]
@DusanJovic-NOAA The total run time 1hr 53min is for 154 nodes or 172nodes?

**Comment by DusanJovic-NOAA** [comment]
> [@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA) The total run time 1hr 53min is for 154 nodes or 172nodes?

154: 1h53m
172: 1h48m
